libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/infopanel/InfoPanelHeaderComponent.kt

```diff
@@ -71,32 +80,28 @@ internal class InfoPanelHeaderComponent(
         }
     }
 
-    private fun updateNavigationStateWhenRouteIsReady(navigationState: NavigationState) {
-        ifNonNull(
+    /**
+     * Dispatch FetchPoints action and wait for RoutesState.Ready.
+     * Method returns immediately if already in RoutesState.Ready or RoutesState.Fetching, or if
+     * required location or destination data is missing.
+     *
+     * @return `true` once in RoutesState.Ready state, otherwise `false`
+     */
+    private suspend fun fetchRouteIfNeeded(): Boolean {
+        if (routesViewModel.state.value is RoutesState.Ready) return true
+        if (routesViewModel.state.value is RoutesState.Fetching) return false
+
+        return ifNonNull(
             locationViewModel.lastPoint,
             destinationViewModel.state.value.destination
         ) { lastPoint, destination ->
             routesViewModel.invoke(RoutesAction.FetchPoints(listOf(lastPoint, destination.point)))
-            coroutineScope.launch {
-                // Wait for fetching to complete and then take action
-                val isRouteReady = waitForFetched()
-                if (isActive && isRouteReady) {
-                    navigationStateViewModel.invoke(
-                        NavigationStateAction.Update(navigationState)
-                    )
-                } else {
-                    logW(TAG, "Routes are not ready")
-                }
-            }
-        } ?: logE(TAG, "Cannot fetch routes because state is incorrect")
+            waitForFetched()
+        } ?: false
     }
 
     private suspend fun waitForFetched(): Boolean {
```

Comment:
nitpick: Just noticed this, seems better to say waitForReady. Up to you

```suggestion
    private suspend fun waitForReady(): Boolean {
```

Reply:
Done